pallets/proxy-keystore/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+
+            // Check Keys storage.
+            for (key_id, key) in keys_iter {
+                if key_id.0 == add_key.key {
+                    // Key with the same hash exists.
+                    return Err(Error::<T>::KeyWithHashExists.into())
+                }
+
+                if key_id.1 != add_key.purpose {
+                    continue
+                }
+
+                // Key with the same purpose was found, check if it's revoked.
+                if let None = key.revoked_at {
+                    // Non-revoked key with the same purpose exists.
+                    return Err(Error::<T>::KeyWithPurposeExists.into())
+                }
+            }
+
+            let key_id: KeyId<T::Hash> = (add_key.key.clone(), add_key.purpose.clone());
+
+            // Check LastKeyByPurpose storage.
+            <LastKeyByPurpose<T>>::try_mutate(
+                account_id.clone(),
+                add_key.purpose.clone(),
+                |key_id_opt| -> DispatchResult {
+                    match key_id_opt {
+                        // Last key by purpose found.
+                        Some(old_key_id) => {
+                            // Extra check to ensure that we don't have any invalid keys in here.
+                            if old_key_id.0 == add_key.key {
+                                return Err(Error::<T>::KeyWithHashExists.into())
+                            }
+                        },
+                        // No last key by purpose, we can continue.
+                        None => {},
+                    }
+
+                    // Replace any value that we might have in there since checks were OK.
+                    let _ = key_id_opt.insert(key_id.clone());
+
+                    Ok(())
+                },
+            )?;
+
+            // Insert the new key.
+            <Keys<T>>::insert(
+                account_id.clone(),
+                key_id,
+                Key{
+                    purpose: add_key.purpose.clone(),
+                    key_type: add_key.key_type.clone(),
+                    revoked_at: None,
+                    deposit: key_deposit,
+                },
+            );
+
+            Ok(())
+        }
+
+        /// Revoke a key with `key_hash` at the current `block_number` in the `Keys` storage
+        /// if the key is found and *not* revoked.
+        ///
+        /// Any entry that matches the `key_hash` and the key purpose in the `LastKeyByPurpose`
+        /// storage is removed.
+        ///
+        /// The key deposit is returned upon success.
+        fn revoke_key(
+            account_id: T::AccountId,
+            key_hash: T::Hash,
+            block_number: T::BlockNumber,
+        ) -> DispatchResult {
+            let mut key_id_opt: Option<KeyId<T::Hash>> = None;
+
+            let iter = <Keys<T>>::iter_prefix(account_id.clone());
+
+            // Search for a key that has the key_hash that we are looking for.
+            for (key_id, _) in iter {
```

Comment:
Wouldn't it make sense to iterate over the supported purposes enum instead? 
```
for kp in KeyPurpose {
   <Keys<T>>::try_mutate(account_id, key_id(key, kp)) .. |key_opt| {
      ...
   }
}
```

That way we only direct reference to the key that we care about. And the for loop is known always to be 2. In the future if we add more key purposes it will increment but still known size, and if we remove some we will in any case run a migration.

Reply:
That's a nice idea, will definitely do that. Cheers!